# HW3 - Stock Movement Prediction

作業檔案：
- hw3.ipynb

資料：
https://www.sharecast.com/index/SP_500/prices/download

- train.csv: S&P 500 訓練資料(2009-2017)
- test.csv: S&P 500 測試資料(2018)


In [1]:
# Read data

import pandas as pd
import numpy as np

train_data_path = './train.csv'
test_data_path = './test.csv'

train_df = pd.read_csv(train_data_path)
test_df = pd.read_csv(test_data_path)

print(train_df.shape)
print(train_df.head())
print(test_df.shape)
print(test_df.head())

(2264, 6)
          Date  Open Price  Close Price  High Price  Low Price      Volume
0  02-Jan-2009      902.99       931.80      934.73     899.35  4048270080
1  05-Jan-2009      929.17       927.45      936.63     919.53  5413910016
2  06-Jan-2009      931.17       934.70      943.85     927.28  5392620032
3  07-Jan-2009      927.45       906.65      927.45     902.37  4704940032
4  08-Jan-2009      905.73       909.73      910.00     896.81  4991549952
(252, 6)
          Date  Open Price  Close Price  High Price  Low Price      Volume
0  02-Jan-2018     2683.73      2695.81     2695.89    2682.36  1846463232
1  03-Jan-2018     2697.85      2713.06     2714.37    2697.77  2090595328
2  04-Jan-2018     2719.31      2723.99     2729.29    2719.07  2100767744
3  05-Jan-2018     2731.33      2743.15     2743.45    2727.92  1918869120
4  08-Jan-2018     2742.67      2747.71     2748.51    2737.60  1894823936


In [2]:
# Drop unnecessary columns

drop_col_names = ['Date'] # !--- or you can modify it to drop the columns you don't want ---!

train_df.drop(columns=drop_col_names, inplace=True)
test_df.drop(columns=drop_col_names, inplace=True)

print(train_df.shape)
print(train_df.head())
print(test_df.shape)
print(test_df.head())

(2264, 5)
   Open Price  Close Price  High Price  Low Price      Volume
0      902.99       931.80      934.73     899.35  4048270080
1      929.17       927.45      936.63     919.53  5413910016
2      931.17       934.70      943.85     927.28  5392620032
3      927.45       906.65      927.45     902.37  4704940032
4      905.73       909.73      910.00     896.81  4991549952
(252, 5)
   Open Price  Close Price  High Price  Low Price      Volume
0     2683.73      2695.81     2695.89    2682.36  1846463232
1     2697.85      2713.06     2714.37    2697.77  2090595328
2     2719.31      2723.99     2729.29    2719.07  2100767744
3     2731.33      2743.15     2743.45    2727.92  1918869120
4     2742.67      2747.71     2748.51    2737.60  1894823936


In [3]:
# Add the column `Tomorrow Movement` by comparing the `Close Price` with the previous days as the training target

train_df['Tomorrow Movement'] = np.where(train_df['Close Price'].diff() >= 0, 1, 0)
test_df['Tomorrow Movement'] = np.where(test_df['Close Price'].diff() >= 0, 1, 0)

train_df['Tomorrow Movement'] = train_df['Tomorrow Movement'].shift(-1)
test_df['Tomorrow Movement'] = test_df['Tomorrow Movement'].shift(-1)

print(train_df.head())
print(train_df.tail())

   Open Price  Close Price  High Price  Low Price      Volume  \
0      902.99       931.80      934.73     899.35  4048270080   
1      929.17       927.45      936.63     919.53  5413910016   
2      931.17       934.70      943.85     927.28  5392620032   
3      927.45       906.65      927.45     902.37  4704940032   
4      905.73       909.73      910.00     896.81  4991549952   

   Tomorrow Movement  
0                0.0  
1                1.0  
2                0.0  
3                1.0  
4                0.0  
      Open Price  Close Price  High Price  Low Price      Volume  \
2259     2684.22      2683.34     2685.35    2678.13  1383888512   
2260     2679.09      2680.50     2682.74    2677.96  1103808384   
2261     2682.10      2682.62     2685.64    2678.91  1149108352   
2262     2686.10      2687.54     2687.66    2682.69  1126089856   
2263     2689.15      2673.61     2692.12    2673.61  1332374016   

      Tomorrow Movement  
2259                0.0  
2260      

In [4]:
# !--- You can add your own data preprocessing here ---!

In [5]:
# Drop rows with NaN values

train_df = train_df.dropna()
test_df = test_df.dropna()

print(train_df.shape)

(2263, 6)


In [6]:
# Divide x and y data

train_x_df = train_df.drop(columns=['Tomorrow Movement'])
train_y_df = train_df['Tomorrow Movement']

test_x_df = test_df.drop(columns=['Tomorrow Movement'])
test_y_df = test_df['Tomorrow Movement']

print(train_x_df.shape)
print(train_x_df.head())
print(train_y_df.shape)
print(train_y_df.head())
print('-----')
print(test_x_df.shape)
print(test_x_df.head())
print(test_y_df.shape)
print(test_y_df.head())

(2263, 5)
   Open Price  Close Price  High Price  Low Price      Volume
0      902.99       931.80      934.73     899.35  4048270080
1      929.17       927.45      936.63     919.53  5413910016
2      931.17       934.70      943.85     927.28  5392620032
3      927.45       906.65      927.45     902.37  4704940032
4      905.73       909.73      910.00     896.81  4991549952
(2263,)
0    0.0
1    1.0
2    0.0
3    1.0
4    0.0
Name: Tomorrow Movement, dtype: float64
-----
(251, 5)
   Open Price  Close Price  High Price  Low Price      Volume
0     2683.73      2695.81     2695.89    2682.36  1846463232
1     2697.85      2713.06     2714.37    2697.77  2090595328
2     2719.31      2723.99     2729.29    2719.07  2100767744
3     2731.33      2743.15     2743.45    2727.92  1918869120
4     2742.67      2747.71     2748.51    2737.60  1894823936
(251,)
0    1.0
1    1.0
2    1.0
3    1.0
4    1.0
Name: Tomorrow Movement, dtype: float64


In [7]:
# Normalize data

# !--- Modify here if you want ---!

from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
scaler.fit(train_x_df)

normalized_train_x_df = scaler.transform(train_x_df)
normalized_train_x_df = np.transpose(normalized_train_x_df)

normalized_train_x_df = pd.DataFrame({
    'Open Price': normalized_train_x_df[0],
    'Close Price': normalized_train_x_df[1],
    'High Price': normalized_train_x_df[2],
    'Low Price': normalized_train_x_df[3],
    'Volume': normalized_train_x_df[4],
})

normalized_test_x_df = scaler.transform(test_x_df)
normalized_test_x_df = np.transpose(normalized_test_x_df)
normalized_test_x_df = pd.DataFrame({
    'Open Price': normalized_test_x_df[0],
    'Close Price': normalized_test_x_df[1],
    'High Price': normalized_test_x_df[2],
    'Low Price': normalized_test_x_df[3],
    'Volume': normalized_test_x_df[4],
})

print(normalized_train_x_df.head())
print(train_y_df[:5])

   Open Price  Close Price  High Price  Low Price    Volume
0   -1.552572    -1.494607   -1.505683  -1.541181  0.813175
1   -1.498571    -1.503581   -1.501760  -1.499581  1.823826
2   -1.494446    -1.488625   -1.486853  -1.483605  1.808070
3   -1.502119    -1.546489   -1.520714  -1.534956  1.299148
4   -1.546921    -1.540136   -1.556744  -1.546417  1.511255
0    0.0
1    1.0
2    0.0
3    1.0
4    0.0
Name: Tomorrow Movement, dtype: float64


# Logistic Regression

In [8]:
# Train & Predict using Logistic Regression

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

#lr_model = LogisticRegression() # !--- Initialize the model here ---!
#lr_model.fit(normalized_train_x_df, train_y_df) # !-- Fill the training data here --!

lr_best_acc = 0
for i in range(1,10):
    lr_model = LogisticRegression(penalty='l1',C=i)
    lr_model.fit(normalized_train_x_df, train_y_df)
    
    lr_training_acc = accuracy_score(train_y_df, lr_model.predict(normalized_train_x_df))
    
    lr_predict_test_result = lr_model.predict(normalized_test_x_df)
    lr_testing_acc = accuracy_score(test_y_df, lr_predict_test_result)
    if lr_testing_acc > lr_best_acc:
        lr_best_train_acc = lr_training_acc
        lr_best_acc = lr_testing_acc

print('training accuracy:')
# !-- Predict training target & print the training accuracy here --!
#lr_training_acc = accuracy_score(train_y_df, lr_model.predict(normalized_train_x_df))
print(lr_best_train_acc)

print('\ntesting accuracy:')
# !-- Predict testing target & print the testing accuracy here --!
#lr_predict_test_result = lr_model.predict(normalized_test_x_df)
#lr_testing_acc = accuracy_score(test_y_df, lr_predict_test_result)
print(lr_best_acc)

print('\npredicted testing labels:')
print(lr_predict_test_result)

c:\users\jaden\appdata\local\programs\python\python37\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
c:\users\jaden\appdata\local\programs\python\python37\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
c:\users\jaden\appdata\local\programs\python\python37\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
c:\users\jaden\appdata\local\programs\python\python37\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
c:\users\jaden\appdata\local\programs\python\python37\lib\si

training accuracy:
0.5470614228899691

testing accuracy:
0.5418326693227091

predicted testing labels:
[1. 1. 1. 1. 1. 1. 1. 1. 0. 1. 0. 1. 1. 0. 1. 1. 1. 0. 1. 1. 1. 1. 1. 1.
 0. 1. 1. 0. 0. 0. 0. 0. 1. 1. 1. 1. 0. 0. 1. 1. 1. 0. 0. 1. 0. 1. 0. 1.
 1. 1. 1. 1. 1. 1. 1. 1. 1. 0. 1. 1. 0. 1. 0. 0. 1. 1. 1. 0. 1. 1. 1. 1.
 1. 1. 1. 1. 1. 1. 1. 0. 1. 1. 0. 1. 0. 0. 1. 1. 0. 0. 1. 1. 1. 1. 1. 1.
 1. 1. 0. 1. 1. 1. 0. 1. 1. 1. 1. 0. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
 1. 1. 1. 0. 1. 0. 1. 0. 0. 0. 1. 1. 0. 1. 1. 0. 1. 1. 1. 1. 1. 0. 1. 1.
 1. 1. 1. 0. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 0. 1.
 1. 1. 1. 1. 1. 1. 0. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
 1. 1. 1. 1. 1. 0. 1. 0. 1. 1. 1. 1. 0. 1. 0. 1. 1. 0. 1. 0. 1. 1. 0. 0.
 1. 1. 1. 1. 1. 0. 0. 1. 1. 1. 1. 0. 0. 0. 1. 0. 1. 1. 1. 0. 1. 0. 1. 1.
 1. 1. 1. 1. 1. 1. 1. 1. 0. 0. 1.]


In [9]:
# Print precision, recall, fbeta-score and confusion matrix

from sklearn.metrics import precision_recall_fscore_support
from sklearn.metrics import confusion_matrix

print('precision, recall, fbeta-score:')
print(precision_recall_fscore_support(test_y_df, lr_predict_test_result, average='weighted'))
print('\nconfusion matrix(tn, fp, fn, tp):')
tn, fp, fn, tp = confusion_matrix(test_y_df, lr_predict_test_result).ravel()
print((tn, fp, fn, tp))

precision, recall, fbeta-score:
(0.490867040313323, 0.5059760956175299, 0.46845659393221956, None)

confusion matrix(tn, fp, fn, tp):
(27, 92, 32, 100)


# SVM

In [10]:
# Train & Predict with SVC

from sklearn.svm import SVC

#svc_model = SVC() # !--- Initialize the model here ---!
#svc_model.fit(normalized_train_x_df, train_y_df) # !-- Fill the training data here --!

svc_best_acc = 0
for i in range(1,10):
    svc_model = SVC(C=i)
    svc_model.fit(normalized_train_x_df, train_y_df)
    
    svc_training_acc = accuracy_score(train_y_df, svc_model.predict(normalized_train_x_df))
    
    svc_predict_test_result = svc_model.predict(normalized_test_x_df)
    svc_testing_acc = accuracy_score(test_y_df, svc_predict_test_result)
    if svc_testing_acc > svc_best_acc:
        svc_best_training_acc = svc_training_acc
        svc_best_acc = svc_testing_acc

print('training accuracy:')
# !-- Predict training target & print the training accuracy here --!
#svc_training_acc = accuracy_score(train_y_df, svc_model.predict(normalized_train_x_df))
print(svc_best_training_acc)

print('\ntesting accuracy:')
# !-- Predict testing target & print the testing accuracy here --!
#svc_predict_test_result = svc_model.predict(normalized_test_x_df)
#svc_testing_acc = accuracy_score(test_y_df, svc_predict_test_result)
print(svc_best_acc)

print('\npredicted testing labels:')
print(svc_predict_test_result)

training accuracy:
0.5528060097216085

testing accuracy:
0.6095617529880478

predicted testing labels:
[1. 1. 1. 1. 1. 1. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 1. 1. 1. 1. 1. 1. 0. 1. 1. 1. 1. 0. 0. 0. 0. 1. 1. 1. 1. 1. 0. 0.
 0. 0. 1. 0. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
 1. 1. 1. 1. 1. 1. 1. 0. 1. 1. 1. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 1. 1. 0.
 0. 0. 0. 0. 0. 0. 0. 1. 1. 1. 1. 1. 1. 0. 1. 0. 0. 0. 0. 0. 1. 1. 1. 1.
 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]


In [11]:
# Print precision, recall, fbeta-score and confusion matrix

print('precision, recall, fbeta-score:')
print(precision_recall_fscore_support(test_y_df, svc_predict_test_result, average='weighted'))
print('\nconfusion matrix(tn, fp, fn, tp):')
tn, fp, fn, tp = confusion_matrix(test_y_df, svc_predict_test_result).ravel()
print((tn, fp, fn, tp))

precision, recall, fbeta-score:
(0.6103152606963451, 0.6055776892430279, 0.60515186218688, None)

confusion matrix(tn, fp, fn, tp):
(78, 41, 58, 74)


# Neural Network

In [12]:
# Define NN output groundtruth

falling_prob = pd.DataFrame(data=np.where(train_y_df == 0, 1, 0)[:])
train_y_df = pd.DataFrame(data=np.where(train_y_df == 0, 0, 1)[:])
train_y_df = pd.concat( [ falling_prob, train_y_df ], axis=1, ignore_index=True )

falling_prob = pd.DataFrame(data=np.where(test_y_df == 0, 1, 0)[:])
test_y_df = pd.DataFrame(data=np.where(test_y_df == 0, 0, 1)[:])
test_y_df = pd.concat( [ falling_prob, test_y_df ], axis=1, ignore_index=True )

print(train_y_df.shape)
print(train_y_df.head())

(2263, 2)
   0  1
0  1  0
1  0  1
2  1  0
3  0  1
4  1  0


In [83]:
# Define NN structure

import torch
import torch.nn.functional as F

# !--- You can modify the NN structure here ---!
class M_NN(torch.nn.Module):
    def __init__(self, D_in, H, D_out):
        super(M_NN, self).__init__()
        self.linear1 = torch.nn.Linear(D_in, H)
        self.linear2 = torch.nn.Linear(H, D_out)

    def forward(self, x):
        h = self.linear1(x)
        acti_out = F.relu(h)
        y_pred = self.linear2(h)
        return y_pred


# N = batch size, D_in = input size, H = hidden size, D_out = output size
N, D_in, H, D_out = 100, 5, 70, 2  # !--- You can modify here ---!

model = M_NN(D_in, H, D_out)
criterion = torch.nn.BCEWithLogitsLoss(reduction='mean') # !--- You can modify here ---!
optimizer = torch.optim.SGD(model.parameters(), lr=5e-3) # !--- You can modify here ---!


# Train NN
# !--- You can modify here ---!

for t in range(500):
    for batch_num in range(N, len(normalized_train_x_df), N): 
        y_pred = model(torch.FloatTensor(normalized_train_x_df[batch_num-N : batch_num].values.tolist())) # !-- Fill the training batch data here --!
        loss = criterion(y_pred, torch.FloatTensor(train_y_df[batch_num-N : batch_num].values.tolist())) # !-- Fill the prediction & groundtruth here to calculate loss --!
        
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
        
    if (t%100 == 0):
        print('epoch:%d - loss:%.10f' % (t, loss.item()))

epoch:0 - loss:0.6970199347
epoch:100 - loss:0.6867566705
epoch:200 - loss:0.6864835620
epoch:300 - loss:0.6864428520
epoch:400 - loss:0.6864324808


In [84]:
# Predict

nn_predict_train_y = model(torch.FloatTensor(normalized_train_x_df.values.tolist()))# !-- Predict training data here --!
result_train = np.where(nn_predict_train_y[:, 0] > nn_predict_train_y[:, 1], 1, 0) # !-- You can modify here --!
print('training accuracy:')
print(accuracy_score(train_y_df[0], result_train))

nn_predict_test_y = model(torch.FloatTensor(normalized_test_x_df.values.tolist())) # !-- Predict training data here --!
result_test = np.where(nn_predict_test_y[:, 0] > nn_predict_test_y[:, 1], 1, 0) # !-- You can modify here --!
print('\ntesting accuracy:')
print(accuracy_score(test_y_df[0], result_test))

print('\npredicted testing prob:')
print(nn_predict_test_y)
print('\npredicted testing labels:')
print(result_test)

training accuracy:
0.5475033141847105

testing accuracy:
0.5258964143426295

predicted testing prob:
tensor([[-0.0616,  0.0670],
        [-0.0564,  0.0639],
        [-0.0579,  0.0615],
        [-0.0560,  0.0611],
        [-0.0582,  0.0598],
        [-0.0576,  0.0592],
        [-0.0590,  0.0593],
        [-0.0508,  0.0577],
        [-0.0472,  0.0552],
        [-0.0637,  0.0562],
        [-0.0446,  0.0532],
        [-0.0536,  0.0521],
        [-0.0464,  0.0496],
        [-0.0395,  0.0508],
        [-0.0452,  0.0474],
        [-0.0512,  0.0482],
        [-0.0504,  0.0471],
        [-0.0338,  0.0458],
        [-0.0515,  0.0469],
        [-0.0523,  0.0492],
        [-0.0510,  0.0480],
        [-0.0439,  0.0498],
        [-0.0764,  0.0584],
        [-0.1089,  0.0782],
        [-0.0328,  0.0693],
        [-0.0620,  0.0683],
        [-0.1238,  0.0843],
        [-0.0760,  0.0780],
        [-0.0617,  0.0729],
        [-0.0649,  0.0727],
        [-0.0470,  0.0700],
        [-0.0597,  0.0642],
   

In [37]:
# Print precision, recall, fbeta-score and confusion matrix

print('\nprecision, recall, fbeta-score:')
print(precision_recall_fscore_support(test_y_df[0], result_test, average='weighted'))
print('\nconfusion matrix(tn, fp, fn, tp):')
tn, fp, fn, tp = confusion_matrix(test_y_df[0], result_test).ravel()
print((tn, fp, fn, tp))


precision, recall, fbeta-score:
(0.2765670386184346, 0.5258964143426295, 0.3624977895207681, None)

confusion matrix(tn, fp, fn, tp):
(132, 0, 119, 0)


# Discussion
在三種方法中(Logistic regression、SVM、Neural network)，我分別調整model內的參數試圖提升其準確率，詳細如下：<br>
1.Logistic Regression:<br>
利用迴圈去跑C=1~10個別的準確率，並選擇當中test data準確率最高者；調整penalty='l1'。其他參數調整未有好效果，最終accuracy小幅上升至0.542左右。


2.SVM:<br>
一樣利用迴圈去跑C=1~10個別的準確率，並選擇當中test data準確率最高者。其他參數調整未有好效果，最終accuracy上升至0.61左右，效果較為顯著。


3.Neural Network:<br>
我嘗試調整了batch size和hidden size的大小、torch.nn.BCEWithLogitsLoss裡reduction改為'mean'、調整torch.optim.SGD裡的learning rate、調整epoch的range。上述方法都無法成功有顯著的提升，最終的accuracy小幅上升至0.526左右。


未來可改善的地方：<br>
1.想辦法更了解每個model中參數的意義，以方便調整合適於資料屬性的參數<br>
2.查找一些與股票預測相關的指數或方法等等，利用增加新的feature去提升預測準確率